### PR TITLE
Bluetooth

### PR DESCRIPTION
**Pull request by linknum23**
Integrates kjk2010's single device Bluetooth support
- [x] Update install script

Feedback from @gorski123:
- [x] Add some useful feedback in the title to indicate playback state (Note: we need to make specific fields for this in the future)
   - "Enter Sandman - MotoX 5"
   - "Connected to MotoX 5: play some music"
   - "No device connected - Pair phone/device to 'amplipi' to connect"
- [ ] Future - Add full advanced Bluetooth interface for adding and removing paired devices

### CODE REVIEW

File: scripts/configure.py [hunk 1]
```diff
@@ -173,6 +173,47 @@
...
+    #           {'from': 'config/bluetooth/bluealsa.service', 'to': '/lib/systemd/system/'},
+    #           {'from': 'streams/bluetooth_agent', 'to': '/usr/local/bin/'},
+    #           {'from': 'config/bluetooth/bluetooth_agent.service', 'to': '/etc/systemd/system/'}],
+    'apt' : [ 'libsndfile1', 'libsndfile1-dev', 'libbluetooth-dev', 'bluealsa', 'python-dbus',
+              'libasound2-dev', 'git', 'autotools-dev', 'automake', 'libtool', 'm4' ],
+    'script' : [
+      # 'sudo cp bin/ARCH/rtl8761b_fw /lib/firmware/rtl_bt/rtl8761b_fw.bin',
+      # 'sudo cp bin/ARCH/rtl8761b_config /lib/firmware/rtl_bt/rtl8761b_config.bin',
+      # TODO: handle ARCH keyword in scripts, not just copy
+      'sudo cp bin/arm/rtl8761b_fw /lib/firmware/rtl_bt/rtl8761b_fw.bin',
```
Comment (linknum23) [inline scripts/configure.py:191]: use the env['bas_dir'] to get the amplipi root directory (typically /home/pi/amplipi-dev/)

File: streams/bluetooth.py [hunk 2]
```diff
@@ -0,0 +1,176 @@
+#! /usr/bin/python3
+# -*- coding: utf-8 -*-
+
+# Uses rtl_fm to access an RTL-SDR dongle, redsea to process RDS data (if any), and aplay to output to audio device
```
Comment (linknum23) [inline streams/bluetooth.py:4]: Probably make a comment about what this script does. The rtl_fm comment is old and should be removed.

Comment (linknum23) [reply streams/bluetooth.py:4]: You may want to run all these python scripts through pylint just to pick up on some weird convention stuff.

File: README.md [hunk 3]
```diff
@@ -29,10 +29,10 @@ Most of these digital streaming services are provided by other open-source proje
 |Plexamp|No|No|No|[Plexamp](https://plexamp.com/)|See [#200](http://github.com/micro-nova/AmpliPi/issues/200)|
 |FM Radio|Yes|No|Yes|[rtl-sdr](https://osmocom.org/projects/rtl-sdr/wiki/Rtl-sdr)/[redsea](https://github.com/windytan/redsea)|Requires [RTL SDR](https://www.rtl-sdr.com/buy-rtl-sdr-dvb-t-dongles/) compatible USB device|
 |LMS|Yes|Yes|No|[squeezelite](https://github.com/ralph-irving/squeezelite)|Requires separate LMS Server|
+|Bluetooth|Yes|No|Yes|[bluezero](https://github.com/ukBaz/python-bluezero)||
```
Comment (linknum23) [inline README.md:32]: Add a note linking to the bluetooth.md readme
```suggestion
|Bluetooth|Yes|No|Yes|[bluezero](https://github.com/ukBaz/python-bluezero)|See [Bluetooth docs](docs/BLUETOOTH.md)|
```

File: amplipi/ctrl.py [hunk 4]
```diff
@@ -274,6 +275,27 @@ def reinit(self, settings: models.AppSettings = models.AppSettings(), change_not
...
+          self.delete_stream(s, internal=True)
+    else:
+      print('bluetooth dongle unavailable')
+      if len(bt_streams) > 0:
+        print('bt streams present. removing all')
+        for s in bt_streams:
+          self.delete_stream(s, internal=True)
+
+
+
```
Comment (linknum23) [inline amplipi/ctrl.py:298]:
```suggestion
```

File: scripts/configure.py [hunk 5]
```diff
@@ -173,6 +173,47 @@
...
+      'git clone https://git.kernel.org/pub/scm/bluetooth/sbc.git',
+      'cd sbc',
+      'git checkout 8dc5d5ba381512ad5b1afa45c63ec6b0a3833244',  # sbc release 2.0
+      'sudo ./bootstrap-configure',
+      'sudo ./configure',
+      'sudo make',
+      'sudo make install',
+      'popd',
+      'else',
+      'echo SBC already installed, skipping installation.',
```
Comment (linknum23) [inline scripts/configure.py:205]: Indent please

**Review by linknum23 — COMMENTED** [review]
Some initial comments before hands on testing

**Comment by linknum23** [comment]
Running this on AmpliPi13 the install failed:
```
run bluetooth install script : [['bash', '/home/pi/amplipi-dev/install_bluetooth.sh']]
  cp: cannot stat 'config/bluetooth/main.conf': No such file or directory
  cp: cannot stat 'config/bluetooth/bluealsa.service': No such file or directory
  cp: cannot stat 'config/bluetooth/bluetooth_agent.service': No such file or directory
  Installing SBC...
  /tmp/tmp.tjec1J6ogP ~/amplipi-dev
  Cloning into 'sbc'...
  Note: checking out '8dc5d5ba381512ad5b1afa45c63ec6b0a3833244'.
  You are in 'detached HEAD' state. You can look around, make experimental
  changes and commit them, and you can discard any commits you make in this
  state without impacting any branches by performing another checkout.
  If you want to create a new branch to retain commits you create, you may
  do so (now or later) by using -b with the checkout command again. Example:
    git checkout -b <new-branch-name>
  HEAD is now at 8dc5d5b Release 2.0
  configure.ac:18: installing './compile'
  configure.ac:23: installing './config.guess'
  configure.ac:23: installing './config.sub'
  configure.ac:4: installing './install-sh'
  configure.ac:4: installing './missing'
  Makefile.am: installing './depcomp'
  checking for a BSD-compatible install... /usr/bin/install -c
  checking whether build environment is sane... yes
  checking for a thread-safe mkdir -p... /bin/mkdir -p
  checking for gawk... no
  checking for mawk... mawk
  checking whether make sets $(MAKE)... yes
  checking whether make supports nested variables... yes
  checking how to create a pax tar archive... gnutar
  checking whether make supports nested variables... (cached) yes
  checking whether to enable maintainer-specific portions of Makefiles... yes
  checking for gcc... gcc
  checking whether the C compiler works... yes
  checking for C compiler default output file name... a.out
  checking for suffix of executables... 
  checking whether we are cross compiling... no
  checking for suffix of object files... o
  checking whether we are using the GNU C compiler... yes
  checking whether gcc accepts -g... yes
  checking for gcc option to accept ISO C89... none needed
  checking whether gcc understands -c and -o together... yes
  checking whether make supports the include directive... yes (GNU style)
  checking dependency style of gcc... gcc3
  checking whether gcc accepts -fPIE... yes
  checking build system type... armv7l-unknown-linux-gnueabihf
  checking host system type... armv7l-unknown-linux-gnueabihf
  checking how to print strings... printf
  checking for a sed that does not truncate output... /bin/sed
  checking for grep that handles long lines and -e... /bin/grep
  checking for egrep... /bin/grep -E
  checking for fgrep... /bin/grep -F
  checking for ld used by gcc... /usr/bin/ld
  checking if the linker (/usr/bin/ld) is GNU ld... yes
  checking for BSD- or MS-compatible name lister (nm)... /usr/bin/nm -B
  checking the name lister (/usr/bin/nm -B) interface... BSD nm
  checking whether ln -s works... yes
  checking the maximum length of command line arguments... 1572864
  checking how to convert armv7l-unknown-linux-gnueabihf file names to armv7l-unknown-linux-gnueabihf format... func_convert_file_noop
  checking how to convert armv7l-unknown-linux-gnueabihf file names to toolchain format... func_convert_file_noop
  checking for /usr/bin/ld option to reload object files... -r
  checking for objdump... objdump
  checking how to recognize dependent libraries... pass_all
  checking for dlltool... no
  checking how to associate runtime and link libraries... printf %s\n
  checking for ar... ar
  checking for archiver @FILE support... @
  checking for strip... strip
  checking for ranlib... ranlib
  checking command to parse /usr/bin/nm -B output from gcc object... ok
  checking for sysroot... no
  checking for a working dd... /bin/dd
  checking how to truncate binary pipes... /bin/dd bs=4096 count=1
  checking for mt... mt
  checking if mt is a manifest tool... no
  checking how to run the C preprocessor... gcc -E
  checking for ANSI C header files... yes
  checking for sys/types.h... yes
  checking for sys/stat.h... yes
  checking for stdlib.h... yes
  checking for string.h... yes
  checking for memory.h... yes
  checking for strings.h... yes
  checking for inttypes.h... yes
  checking for stdint.h... yes
  checking for unistd.h... yes
  checking for dlfcn.h... yes
  checking for objdir... .libs
  checking if gcc supports -fno-rtti -fno-exceptions... no
  checking for gcc option to produce PIC... -fPIC -DPIC
  checking if gcc PIC flag -fPIC -DPIC works... yes
  checking if gcc static flag -static works... yes
  checking if gcc supports -c -o file.o... yes
  checking if gcc supports -c -o file.o... (cached) yes
  checking whether the gcc linker (/usr/bin/ld) supports shared libraries... yes
  checking whether -lc should be explicitly linked in... no
  checking dynamic linker characteristics... GNU/Linux ld.so
  checking how to hardcode library paths into programs... immediate
  checking whether stripping libraries is possible... yes
  checking if libtool supports shared libraries... yes
  checking whether to build shared libraries... yes
  checking whether to build static libraries... no
  checking for pkg-config... /usr/bin/pkg-config
  checking pkg-config is at least version 0.9.0... yes
  checking for SNDFILE... yes
  checking that generated files are newer than configure... done
  configure: creating ./config.status
  config.status: creating Makefile
  config.status: creating sbc/sbc.pc
  config.status: creating config.h
  config.status: executing depfiles commands
  config.status: executing libtool commands
  	sbc 2.0
  checking for a BSD-compatible install... /usr/bin/install -c
  checking whether build environment is sane... yes
  checking for a thread-safe mkdir -p... /bin/mkdir -p
  checking for gawk... no
  checking for mawk... mawk
  checking whether make sets $(MAKE)... yes
  checking whether make supports nested variables... yes
  checking how to create a pax tar archive... gnutar
  checking whether make supports nested variables... (cached) yes
  checking whether to enable maintainer-specific portions of Makefiles... no
  checking for gcc... gcc
  checking whether the C compiler works... yes
  checking for C compiler default output file name... a.out
  checking for suffix of executables... 
  checking whether we are cross compiling... no
  checking for suffix of object files... o
  checking whether we are using the GNU C compiler... yes
  checking whether gcc accepts -g... yes
  checking for gcc option to accept ISO C89... none needed
  checking whether gcc understands -c and -o together... yes
  checking whether make supports the include directive... yes (GNU style)
  checking dependency style of gcc... gcc3
  checking whether gcc accepts -fPIE... yes
  checking build system type... armv7l-unknown-linux-gnueabihf
  checking host system type... armv7l-unknown-linux-gnueabihf
  checking how to print strings... printf
  checking for a sed that does not truncate output... /bin/sed
  checking for grep that handles long lines and -e... /bin/grep
  checking for egrep... /bin/grep -E
  checking for fgrep... /bin/grep -F
  checking for ld used by gcc... /usr/bin/ld
  checking if the linker (/usr/bin/ld) is GNU ld... yes
  checking for BSD- or MS-compatible name lister (nm)... /usr/bin/nm -B
  checking the name lister (/usr/bin/nm -B) interface... BSD nm
  checking whether ln -s works... yes
  checking the maximum length of command line arguments... 1572864
  checking how to convert armv7l-unknown-linux-gnueabihf file names to armv7l-unknown-linux-gnueabihf format... func_convert_file_noop
  checking how to convert armv7l-unknown-linux-gnueabihf file names to toolchain format... func_convert_file_noop
  checking for /usr/bin/ld option to reload object files... -r
  checking for objdump... objdump
  checking how to recognize dependent libraries... pass_all
  checking for dlltool... no
  checking how to associate runtime and link libraries... printf %s\n
  checking for ar... ar
  checking for archiver @FILE support... @
  checking for strip... strip
  checking for ranlib... ranlib
  checking command to parse /usr/bin/nm -B output from gcc object... ok
  checking for sysroot... no
  checking for a working dd... /bin/dd
  checking how to truncate binary pipes... /bin/dd bs=4096 count=1
  checking for mt... mt
  checking if mt is a manifest tool... no
  checking how to run the C preprocessor... gcc -E
  checking for ANSI C header files... yes
  checking for sys/types.h... yes
  checking for sys/stat.h... yes
  checking for stdlib.h... yes
  checking for string.h... yes
  checking for memory.h... yes
  checking for strings.h... yes
  checking for inttypes.h... yes
  checking for stdint.h... yes
  checking for unistd.h... yes
  checking for dlfcn.h... yes
  checking for objdir... .libs
  checking if gcc supports -fno-rtti -fno-exceptions... no
  checking for gcc option to produce PIC... -fPIC -DPIC
  checking if gcc PIC flag -fPIC -DPIC works... yes
  checking if gcc static flag -static works... yes
  checking if gcc supports -c -o file.o... yes
  checking if gcc supports -c -o file.o... (cached) yes
  checking whether the gcc linker (/usr/bin/ld) supports shared libraries... yes
  checking whether -lc should be explicitly linked in... no
  checking dynamic linker characteristics... GNU/Linux ld.so
  checking how to hardcode library paths into programs... immediate
  checking whether stripping libraries is possible... yes
  checking if libtool supports shared libraries... yes
  checking whether to build shared libraries... yes
  checking whether to build static libraries... no
  checking for pkg-config... /usr/bin/pkg-config
  checking pkg-config is at least version 0.9.0... yes
  checking for SNDFILE... yes
  checking that generated files are newer than configure... done
  configure: creating ./config.status
  config.status: creating Makefile
  config.status: creating sbc/sbc.pc
  config.status: creating config.h
  config.status: config.h is unchanged
  config.status: executing depfiles commands
  config.status: executing libtool commands
  	sbc 2.0
  make --no-print-directory all-am
    CC       src/sbcinfo.o
    CCLD     src/sbcinfo
    CC       src/sbcdec.o
    CC       sbc/sbc.lo
    CC       sbc/sbc_primitives.lo
    CC       sbc/sbc_primitives_sse.lo
    CC       sbc/sbc_primitives_mmx.lo
    CC       sbc/sbc_primitives_iwmmxt.lo
    CC       sbc/sbc_primitives_neon.lo
    CC       sbc/sbc_primitives_armv6.lo
    CCLD     sbc/libsbc-private.la
  ar: `u' modifier ignored since `D' is the default (see `U')
    CCLD     src/sbcdec
    CC       src/sbcenc.o
    CCLD     src/sbcenc
    CC       src/sbctester.o
    CCLD     src/sbctester
    CCLD     sbc/libsbc.la
   /bin/mkdir -p '/usr/local/lib'
   /bin/bash ./libtool   --mode=install /usr/bin/install -c   sbc/libsbc.la '/usr/local/lib'
  libtool: install: /usr/bin/install -c sbc/.libs/libsbc.so.1.3.1 /usr/local/lib/libsbc.so.1.3.1
  libtool: install: (cd /usr/local/lib && { ln -s -f libsbc.so.1.3.1 libsbc.so.1 || { rm -f libsbc.so.1 && ln -s libsbc.so.1.3.1 libsbc.so.1; }; })
  libtool: install: (cd /usr/local/lib && { ln -s -f libsbc.so.1.3.1 libsbc.so || { rm -f libsbc.so && ln -s libsbc.so.1.3.1 libsbc.so; }; })
  libtool: install: /usr/bin/install -c sbc/.libs/libsbc.lai /usr/local/lib/libsbc.la
  libtool: finish: PATH="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/sbin" ldconfig -n /usr/local/lib
  ----------------------------------------------------------------------
  Libraries have been installed in:
     /usr/local/lib
  If you ever happen to want to link against installed libraries
  in a given directory, LIBDIR, you must either use libtool, and
  specify the full pathname of the library, or use the '-LLIBDIR'
  flag during linking and do at least one of the following:
     - add LIBDIR to the 'LD_LIBRARY_PATH' environment variable
       during execution
     - add LIBDIR to the 'LD_RUN_PATH' environment variable
       during linking
     - use the '-Wl,-rpath -Wl,LIBDIR' linker flag
     - have your system administrator add LIBDIR to '/etc/ld.so.conf'
  See any operating system documentation about shared libraries for
  more information, such as the ld(1) and ld.so(8) manual pages.
  ----------------------------------------------------------------------
   /bin/mkdir -p '/usr/local/bin'
    /bin/bash ./libtool   --mode=install /usr/bin/install -c src/sbcinfo src/sbcdec src/sbcenc '/usr/local/bin'
  libtool: install: /usr/bin/install -c src/sbcinfo /usr/local/bin/sbcinfo
  libtool: install: /usr/bin/install -c src/sbcdec /usr/local/bin/sbcdec
  libtool: install: /usr/bin/install -c src/sbcenc /usr/local/bin/sbcenc
   /bin/mkdir -p '/usr/local/lib/pkgconfig'
   /usr/bin/install -c -m 644 sbc/sbc.pc '/usr/local/lib/pkgconfig'
   /bin/mkdir -p '/usr/local/include/sbc'
   /usr/bin/install -c -m 644 sbc/sbc.h '/usr/local/include/sbc'
  ~/amplipi-dev
  Synchronizing state of bluetooth.service with SysV service script with /lib/systemd/systemd-sysv-install.
  Executing: /lib/systemd/systemd-sysv-install enable bluetooth
  The unit files have no installation config (WantedBy=, RequiredBy=, Also=,
  Alias= settings in the [Install] section, and DefaultInstance= for template
  units). This means they are not meant to be enabled using systemctl.
   
  Possible reasons for having this kind of units are:
  • A unit may be statically enabled by being symlinked from another unit's
    .wants/ or .requires/ directory.
  • A unit's purpose may be to act as a helper for some other unit which has
    a requirement dependency on it.
  • A unit may be started when needed via activation (socket, path, timer,
    D-Bus, udev, scripted systemctl call, ...).
  • In case of template units, the unit is meant to be enabled with some
    instance name specified.
  Failed to enable unit: Unit file bluetooth_agent.service does not exist.
  Error: Task Failed
remove bluetooth temporary script : [['rm', '/home/pi/amplipi-dev/install_bluetooth.sh']]
```

**Comment by linknum23** [comment]
Looks like you may need to do:
`systemctl --user daemon-reload`

After copying in the services

**Comment by linknum23** [comment]
It looks like the bluetoo0th config files were not copied over. You prabably need to add this line to pyproject.toml:

    "config/bluetooth/*",

File: streams/bluetooth.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+    try:
+      with open(args.log, 'a') as f:
+        print(info, file=f)
+    except:
+      print(f'Error writing to logfile: {args.log}')
+      print(info)
+  else:
+    print(info)
+
+if args.song_info:
```
Comment (linknum23) [inline streams/bluetooth.py:32]: What is this?

File: streams/bluetooth.py [hunk 7]
```diff
@@ -0,0 +1,159 @@
+#! /usr/bin/python3
+# -*- coding: utf-8 -*-
+
+""" Bluetooth currently only supports at most one Bluetooth stream. Handles media controls and audio. """
+
+import time
+import json
+import argparse
+import sys
+from bluezero import dbus_tools
```
Comment (linknum23) [inline streams/bluetooth.py:10]: group these on one line?

File: streams/bluetooth.py (hunk 6 above)
Comment (jonahshader) [reply streams/bluetooth.py:32]: I don't know. It doesn't do anything right now, should I remove it?

File: scripts/configure.py [hunk 8]
```diff
@@ -173,6 +173,47 @@
   },
   'spotify' : {
     'copy' : [{'from': 'bin/ARCH/spotifyd', 'to': 'streams/spotifyd'}],
+  },
+  'bluetooth' : {
+    'amplipi_only' : [],
```
Comment (linknum23) [inline scripts/configure.py:178]:
```suggestion
    'amplipi_only' : True,
```

File: streams/bluetooth.py (hunk 6 above)
Comment (jonahshader) [reply streams/bluetooth.py:32]: I removed it, was probably some debug prints.

File: scripts/configure.py [hunk 9]
```diff
@@ -268,6 +309,8 @@ def print_progress(tasks):
   scripts: Dict[str, List[str]] = {}
   for dep in deps:
     install_steps = _os_deps[dep]
+    if 'amplipi_only' in install_steps and not env['is_amplipi']:
```
Comment (linknum23) [inline scripts/configure.py:312]:
```suggestion
    if not env['is_amplipi'] and install_steps.get('amplipi_only', False):
```

**Comment by jonahshader** [comment]
Another idea is changing the stream name to the device name. Right now the stream type is Bluetooth and the stream name is bluetooth so it shows up in the ui as Bluetooth - bluetooth, which kinda sucks. At the very least this will require interprocess communication, but there could be more hurdles with renaming streams programmatically.

**Comment by linknum23** [comment]
> Another idea is changing the stream name to the device name.
This is entirely possible and doesn't have any real system conflicts. It can be changed at any time without real reprocussions besides the need to serialize the json, but might be weird for some users. We can test it and see if it makes sense.

File: amplipi/streams.py [hunk 10]
```diff
@@ -1097,46 +1096,44 @@ def connect(self, src):
...
+    subprocess.run(args='bluetoothctl power on'.split(), preexec_fn=os.setpgrp)
+    subprocess.run(args='bluetoothctl discoverable on'.split(), preexec_fn=os.setpgrp)
 
     # Start audio via bluealsa-aplay (accepting any bluetooth device with all zeros)
-    baplay_args = f'bluealsa-aplay -d {utils.output_device(src)} 00:00:00:00:00:00'
-    self.proc = subprocess.Popen(args=baplay_args.split(), stdin=subprocess.PIPE, stdout=subprocess.PIPE,
-                                 stderr=subprocess.PIPE, preexec_fn=os.setpgrp)
+    # baplay_args = f'bluealsa-aplay -d {utils.output_device(src)} 00:00:00:00:00:00 --single-audio'
+    # testing valid mac (jonah's phone)
+    baplay_args = f'bluealsa-aplay -d {utils.output_device(src)} 78:F2:38:58:0A:13 --single-audio'
```
Comment (linknum23) [inline amplipi/streams.py:1105]: This hardcoded MAC seems fishy

Comment (jonahshader) [reply amplipi/streams.py]: Dead code, removed

**Comment by linknum23** [comment]
Updated this ahead of reviewing this PR in the hopes of making a Bluetooth prerelease soon.

File: amplipi/streams.py [hunk 11]
```diff
@@ -1075,9 +1077,110 @@ def info(self) -> models.SourceInfo:
...
+    try:
+      if cmd in self.supported_cmds and self.src is not None:
+        src_config_folder = f"{utils.get_folder('config')}/srcs/{self.src}"
+        device_info_path = f'{src_config_folder}/btDevice'
+        btcmd_args = f'{sys.executable} {utils.get_folder("streams")}/bluetooth.py --command={cmd} --device-info={device_info_path}'
+        subprocess.run(args=btcmd_args.split(), preexec_fn=os.setpgrp)
+      else:
+        raise NotImplementedError(f'"{cmd}" is either incorrect or not currently supported')
+    except Exception as e:
+      print(f'bluetooth: exception {e}')
```
Comment (linknum23) [inline amplipi/streams.py:1178]: We should probably re-rasie this: somethign like:
```python
      raise RuntimeError(f'Command {cmd} failed to send: {exc}') from e
```

File: streams/bluetooth.py [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+
+      sys.stdout.flush()
+      time.sleep(1)  # might want to reduce this - currently doesn't feel very responsive
+
+  # TODO: should we sys.exit on these excepts?
+  except KeyboardInterrupt:
+    print('exit')
+
+  except Exception:
+    log(sys.exc_info())
```
Comment (linknum23) [inline streams/bluetooth.py:254]: This looks like it reports the exception and traceback?

**Comment by linknum23** [comment]
@jonahshader this looks really close! Can you make some of the suggested changes so we can test and make a prerelease from there.

File: amplipi/streams.py (hunk 11 above)
Comment (jonahshader) [reply amplipi/streams.py:1178]: What is {exc} here?

**Pull request merged**